### ISSUE TITLE
Indexing crashed with UnexpectedEof and OversizedVectorAllocation

### ISSUE THREAD

**Issue by chautnguyen4392** [rust-litecoin/rust-litecoin#6]
I tried using [electrs-ltc](https://github.com/rust-litecoin/electrs-ltc) (commit: 55d81ebacd3977a98c11de08f5c6dd838c1babe1) to index litecoin blockchain. I used following command
```
RUST_BACKTRACE=full cargo +1.70.0 run --release --bin electrs -- -vvvv --timestamp --daemon-dir "/media/LTCTemp1T" --db-dir "/media/LTCTemp1T/electrs_db" --cookie "mempool:mempool" --http-addr "127.0.0.1:3001" --monitoring-addr "127.0.0.1:4225" --cors "*" --utxos-limit 1000
```
But it crashed with following log and stacktrace
```
2024-05-01T00:33:00.311-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00000.dat"
2024-05-01T00:33:00.372-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00001.dat"
2024-05-01T00:33:00.372-04:00 - TRACE - parsing 134014007 bytes
2024-05-01T00:33:00.650-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00002.dat"
2024-05-01T00:33:00.650-04:00 - TRACE - parsing 134023563 bytes
2024-05-01T00:33:00.689-04:00 - TRACE - fetched 33411 blocks
2024-05-01T00:33:01.639-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00003.dat"
2024-05-01T00:33:01.758-04:00 - TRACE - parsing 134187262 bytes
2024-05-01T00:33:01.791-04:00 - TRACE - fetched 2986 blocks
2024-05-01T00:33:02.533-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00004.dat"
2024-05-01T00:33:02.851-04:00 - DEBUG - writing 3229522 rows to RocksDB { path: "/media/LTCTemp1T/electrs_db/mainnet/newindex/txstore" }, flush=Disable
2024-05-01T00:33:02.868-04:00 - TRACE - parsing 134214450 bytes
2024-05-01T00:33:02.937-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00005.dat"
2024-05-01T00:33:02.994-04:00 - TRACE - fetched 39159 blocks
2024-05-01T00:33:03.194-04:00 - TRACE - parsing 134027371 bytes
2024-05-01T00:33:03.348-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00006.dat"
2024-05-01T00:33:05.150-04:00 - TRACE - parsing 134212386 bytes
2024-05-01T00:33:05.150-04:00 - TRACE - reading "/media/LTCTemp1T/blocks/blk00007.dat"
thread 'parse-blocks-1' panicked at 'failed to parse Block: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })', src/new_index/fetch.rs:228:54
stack backtrace:
   0:     0x55f009945e3c - std::backtrace_rs::backtrace::libunwind::trace::h9a6b80bbf328ba5d
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55f009945e3c - std::backtrace_rs::backtrace::trace_unsynchronized::hd162ec543a11886b
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55f009945e3c - std::sys_common::backtrace::_print_fmt::h78a5099be12f51a6
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x55f009945e3c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha1c5390454d74f71
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55f00975813f - core::fmt::write::h9ffde816c577717b
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/fmt/mod.rs:1254:17
   5:     0x55f00991ca64 - std::io::Write::write_fmt::h88186074961638e4
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/io/mod.rs:1698:15
   6:     0x55f00994744f - std::sys_common::backtrace::_print::h184198273ed08d59
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x55f00994744f - std::sys_common::backtrace::print::h1b4d8e7add699453
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x55f00994704e - std::panicking::default_hook::{{closure}}::h393bcea75423915a
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:269:22
   9:     0x55f0099480ae - std::panicking::default_hook::h48c64f31d8b3fd03
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:288:9
  10:     0x55f0099480ae - std::panicking::rust_panic_with_hook::hafdc493a79370062
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:691:13
  11:     0x55f009947ba4 - std::panicking::begin_panic_handler::{{closure}}::h0a64bc82e36bedc7
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:582:13
  12:     0x55f009947b06 - std::sys_common::backtrace::__rust_end_short_backtrace::hc203444fb7416a16
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys_common/backtrace.rs:150:18
  13:     0x55f009947af1 - rust_begin_unwind
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
  14:     0x55f00966f2c2 - core::panicking::panic_fmt::h0f6ef0178afce4f2
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
  15:     0x55f00966f652 - core::result::unwrap_failed::h8090202169109f9c
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/result.rs:1687:5
  16:     0x55f009809b99 - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  17:     0x55f009809d9f - rayon_core::join::join_context::{{closure}}::hff58d78ed39fb891
  18:     0x55f009809a6d - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  19:     0x55f009809d9f - rayon_core::join::join_context::{{closure}}::hff58d78ed39fb891
  20:     0x55f009809a6d - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  21:     0x55f009809d9f - rayon_core::join::join_context::{{closure}}::hff58d78ed39fb891
  22:     0x55f009809a6d - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  23:     0x55f009809d9f - rayon_core::join::join_context::{{closure}}::hff58d78ed39fb891
  24:     0x55f009809a6d - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  25:     0x55f009809d9f - rayon_core::join::join_context::{{closure}}::hff58d78ed39fb891
  26:     0x55f009809a6d - rayon::iter::plumbing::bridge_producer_consumer::helper::h955de4d7fddc7895
  27:     0x55f00980a057 - <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute::h9df475f0fe2f3b10
  28:     0x55f0096963d9 - rayon_core::registry::WorkerThread::wait_until_cold::h2342a19c1094ae2f
  29:     0x55f00990524e - std::sys_common::backtrace::__rust_begin_short_backtrace::h0607650116d989d2
  30:     0x55f009904bf4 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h4836e4ab95c17725
  31:     0x55f009948e15 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha1f2224656a778fb
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/boxed.rs:1973:9
  32:     0x55f009948e15 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::haa29ed9703f354b7
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/boxed.rs:1973:9
  33:     0x55f009948e15 - std::sys::unix::thread::Thread::new::thread_start::h33b6dae3e3692197
                               at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/sys/unix/thread.rs:108:17
2024-05-01T00:33:05.225-04:00 - TRACE - fetched 26912 blocks
thread 'parse-blocks-2' panicked at 'failed to parse Block: OversizedVectorAllocation { requested: 5815438783173361663, max: 4000000 }', src/new_index/fetch.rs:228:54
thread 'parse-blocks-0' panicked at 'failed to parse Block: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })', src/new_index/fetch.rs:228:54
  34:     0x154a45dcf609 - start_thread
                               at /build/glibc-e2p3jK/glibc-2.31/nptl/pthread_create.c:477:8
  35:     0x154a45b9d353 - clone
                               at /build/glibc-e2p3jK/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
  36:                0x0 - <unknown>
stack backtrace:
   0: Aborted (core dumped)
```

Can you help to take a look at this issue ?
Thanks

**Comment by knocte** [comment]
hi @chautnguyen4392 thanks for reporting this bug. We'll look at it next week.

**Comment by chautnguyen4392** [comment]
Thanks @knocte . Let me know if you find anything or need more info from me.

**Comment by webwarrior-ws** [comment]
@chautnguyen4392 can you run electrs-ltc from [this branch](https://github.com/webwarrior-ws/electrs-ltc/commits/debug-block-parsing/) (that adds logging of which blocks are being decoded) and post execution log? Maybe not the whole log, just final portion before crash.

**Comment by chautnguyen4392** [comment]
@webwarrior-ws I run code from your branch, below is the final portion before crash
```
Decoding block 235cfd76abe3831e68506f03a75e8538f52e87aa3d646439be1908d13a53e2d5
  35: Decoding block 7f812f9c3a844fd23fd1cd4571cadcab1d80400b49c4f0d4dfa9d1d806ad2614
   Decoding block 7e8523d6b8ba44434527700b101d3d226916b493e19bcc08023cc8c574ad6e80
 0x7ff2fc684353Decoding block a62760e33140debc428812ce7c3c1c18bf7011c026a5a74033b3d4bef3f1ba02
 - clone
Decoding block cecd91087dec5d948ce208c143330f8c89eb67faaf3d6a7825d3926d2dccb39d
  Decoding block 6a53815fdd971ca632e42d918a29d2aaefc349e852f5b0f33748aeb3523e76d2
  Decoding block 5f64de4669dc6fa03a6d9becf49af6e8cf5d2b9e899736ec9a5ea457f5bad599
  Decoding block 96f002972b211e75139f1dffbff1e3911035f355a74cc86c6516f40ddcb5feab
     Decoding block 9535e255d5049440570242988299c62b10a315cd2caea4c6fc5eeaa9a88135c2
   Decoding block a87017f792ca02f43d01799afa320de71d3749ea1c58753bdc2ca63a97ab77ea
    Decoding block 6737ff06f05f6a3d70aa026a08815945b4fe11d4aa9ad016a4abcadbf8b15bc1
             at Decoding block 7027b96eb6e5d36cd629f3bf7f950371d7d366f2375a3aec9f0ddaeaad485946
/build/glibc-e2p3jK/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95Decoding block 126135b71fa16f10460d81ca4794eb13f11ada0b88281cfbec5204eabb644e30

Decoding block 63ef1b74b046dd2dce14e31f19154269229ca7fe3964cd67cc9c596cea71812b
  36Decoding block fd33d64cb6e79ad99fe8cfee91dc215ca4103ffbc03d24b0d8685a3ce3735e45
: Decoding block 19ee0f1e84b762d20918c348d2034d7b8ff483d38b43d2ade2debf450f23e0b0
   Decoding block 521c27619fe30f263db9b959626e83e7e816f189b4b312bf41c3df65836733a1
  Decoding block 47a18129a396d25dd86f954caf77d08a0f881a93afe2f27feea0dc11fe47e1ca
   Decoding block 5a5c22711b2a9955ad5ef94323e87d7546fade6e475741a4f1427ec41403590d
  Decoding block 0215f91961d3de909fda90fb95473d3ea5ccd0847bc01fbd1ba5c830d9095e3e
  Decoding block 8ce7b21682591b389043ccae95b2e6f9dea3ab839ab95a139f75a44aa06e7fca
   0xDecoding block c90db5cbc6fb03145920d9e9f0c3c4d818d8f79f1c24b1aaae8b9fa49323002f
0 - <unknown>
Decoding block 8d0bd7d3e725dbd4c273b2f0996d0b19d04038eb0c5f382bce0188b4c52c724f
Decoding block 727f5fad3e9abf81f58bd8542ec766a701b09b790a84d785e1f0cedb1eff7780
Aborted (core dumped)
```
For your reference, I also attach  [full.log](https://github.com/rust-litecoin/rust-litecoin/files/15307106/full.log)

**Comment by webwarrior-ws** [comment]
@chautnguyen4392 Unfortunately since electrs-ltc processes blocks in parallel, that log output wasn't helpful in determining which block caused failure. I updated my electrs-ltc and rust-litecoin debug branches to only log on parsing failure. Can you run once more?

**Comment by chautnguyen4392** [comment]
@webwarrior-ws I ran your latest code, below is the block causing error
```
2024-05-21T09:26:02.560-04:00 - DEBUG - writing 0 rows to RocksDB { path: "/media/LTCDataNew/electrs_db/mainnet/newindex/txstore" }, flush=Disable
2024-05-21T09:26:02.560-04:00 - TRACE - parsing 134212386 bytes
2024-05-21T09:26:02.560-04:00 - TRACE - reading "/media/LTCDataNew/blocks/blk00007.dat"
Error decoding transactions of block 4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91
Failed to parse block from slice [
```
For your reference, I also attach [full.log](https://github.com/rust-litecoin/rust-litecoin/files/15390519/full_21May2024.log)

**Comment by webwarrior-ws** [comment]
@chautnguyen4392 I tried to create a regression test that parses failed block (`4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91`), but it parses correctly without error.
So I could not reproduce the bug.

**Comment by knocte** [comment]
@chautnguyen4392 maybe you can contribute a regression test that demonstrates the failure in GitHubActions CI? Then we can work on it more easily.

**Comment by webwarrior-ws** [comment]
> @chautnguyen4392 I tried to create a regression test that parses failed block (`4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91`), but it parses correctly without error. So I could not reproduce the bug.

https://github.com/webwarrior-ws/rust-litecoin/commit/87f56437e4e5818aa90816cda7fb6703072c7e95

**Comment by chautnguyen4392** [comment]
@knocte @webwarrior-ws Thanks for your info. I understood that you can't reproduce this issue on your own. This info as well as the info about the block [4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91](https://litecoinspace.org/block/4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91) causing the parse error (thanks to the @webwarrior-ws's debug branch) shed more light for me. I tried using litecoin-cli to get data of this block, and I see this error:
```
$ ./litecoin-cli -rpcuser=mempool -rpcpassword=mempool getblock "4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91" 0
error code: -1
error message:
Block not found on disk
```

There is no problem when I try getting data of blocks before and after the block `4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91`
```
$ ./litecoin-cli -rpcuser=mempool -rpcpassword=mempool getblock "23821d1b5ec7bf4c693e77311257ee0b293a6b68d3b0f6f401a58ae9b09ff1c8" 0
010000005d6132b911b69d1fdb855d2e1c794dfab31a11038793724384775ccc8593303e751c7888d6bbc5e2629a0e53d007d96dc826f344a1474aae9003989727dd85161292de50fb65061c001a17850101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0c0365636f041192de5002b614ffffffff0100f2052a010000001976a9148d21d00a25516436aa5f01fd297dd959155fe9ba88ac00000000
```
```
$ ./litecoin-cli -rpcuser=mempool -rpcpassword=mempool getblock "4773ea7e740e1d510a93d394f73e766459cebee41807014b0a47958480b8f023" 0
01000000915e047064e2d59c4a1349de27efc88f46c4da6053ab833a89279da49280774df62100eaafa14998a8691fbb09640619dc175f584878019b881c37ff52e91ca4fe94de50fb65061c8000042e0301000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0f04fe94de5002b117062f503253482fffffffff0180889e2a01000000434104985534168c7994275ab3d6ded775c2449cf361c50672b5cdd083fd64cbdb8e533131536a8faf5cf9fbe2607c76cfadc78dd4772172b3c3f2ac3201838e51abe1ac000000000100000001ded21536591fbc557eb513b14c6c5ecf76b6d1542070e36cb0e9dc8a909e1546000000004a493046022100dffbdd67e70f4310e8bbe4f19ed327ed5723665c7c57a8e9ca7351817f64c402022100c43b39bb90753be56f82be2478ac7407d68c9ee730ab384713d9a4a2b32e153201ffffffff0248e44e00000000001976a91480f1d0d25e1e806cc81cccdcbaf644abad64b7e888ac38a44f2a010000001976a91452b5de00f0f429b45fd1cbd2120c29a89151847488ac000000000100000001a3ed574239210d89383d42426c4ae2b48772dbe3bdd5f5cedb09f7bf3610b71c000000008c493046022100dee50fb8a01a950ddba0e2427d293be5d9499c3448ba95e1f6bdbea783b8e890022100e3c3fae63b7285afd708ab746a0b5ae78a8fc72b2d8b5298570f39cecaeaf483014104ba5981ffc6a96ba5a7982f52e7f6116bdf6786dc5ccc1830b7e7ef754fc2c42e2186f75a0f35ffb3c56b9297c6254ce7f09e8e99b0fe0eb232a6bc9a9ea74f44ffffffff0228594c01000000001976a9146ba7206fac2b6280a908d74bcf85944c60a4d44b88acd6096006000000001976a9143b4958410293c79ff3b3976b1d977c0ae72b0b3688ac00000000
```

So I think that the litecoin data on my hard disk was corrupted. I'm trying syncing the litecoin data from beginning, then try running [electrs-ltc](https://github.com/rust-litecoin/electrs-ltc) again. Will let you know the result.

**Comment by knocte** [comment]
So litecoin-core doesn't have a checksum of some sort to detect if a block became corrupted?

**Comment by chautnguyen4392** [comment]
@knocte @webwarrior-ws I tried syncing the litecoin data from beginning, then try running [electrs-ltc](https://github.com/rust-litecoin/electrs-ltc) again. It worked properly this time, there is no problem anymore. It means that the old litecoin data stored in my old disk was corrupted. Thanks for your help. I'm really appreciate.
> So litecoin-core doesn't have a checksum of some sort to detect if a block became corrupted?

I think that at the time the litecoind get this block from other peers and store to my old disk, its data is correct. I copied this data from my old disk to a new disk which have better performance. It can be that my old disk have some bad blocks and cause corrupted data or the copy process has issue, so the block `4d778092a49d27893a83ab5360dac4468fc8ef27de49134a9cd5e26470045e91` is corrupted.

@webwarrior-ws Anyway, it's good to add extra debug log to `electrs-ltc` to show the block hash when there is a parsing error (same as what you did). It's really helpful to debug issues
